crates/matrix-sdk/src/widget/matrix.rs

```diff
@@ -309,38 +318,137 @@ impl MatrixDriver {
+            // TODO! this will later be blocked at negotiation level
+            return Ok(Default::default());
+        }
+
         let client = self.room.client();
 
-        let request = if encrypted {
-            return Err(Error::UnknownError(
-                "Sending encrypted to-device events is not supported by the widget driver.".into(),
-            ));
-        } else {
-            RumaToDeviceRequest::new_raw(event_type, TransactionId::new(), messages)
-        };
+        let mut failures: BTreeMap<OwnedUserId, Vec<OwnedDeviceId>> = BTreeMap::new();
 
-        let response = client.send(request).await;
+        let room_encrypted = self
+            .room
+            .latest_encryption_state()
+            .await
+            .map(|s| s.is_encrypted())
+            // Default consider encrypted
+            .unwrap_or(true);
+
+        if room_encrypted {
+            trace!("Sending to-device message in encrypted room <{}>", self.room.room_id());
+
+            //  Attempt to re-batch these by content to batch them up into a single request
+            let mut content_grouped: BTreeMap<
+                &str,
+                BTreeMap<OwnedUserId, Vec<DeviceIdOrAllDevices>>,
+            > = BTreeMap::new();
+
+            for (user_id, device_map) in messages.iter() {
+                for (device_id, content) in device_map.iter() {
+                    content_grouped
+                        .entry(content.json().get())
+                        .or_default()
+                        .entry(user_id.clone())
+                        .or_default()
+                        .push(device_id.to_owned());
+                }
+            }
 
-        response.map_err(Into::into)
+            // Encrypt and send
+            for (content, user_map) in content_grouped {
+                // let's get all the devices for this content
+                let mut device_set: BTreeSet<(OwnedUserId, OwnedDeviceId)> = BTreeSet::new();
+                for (user_id, device_id_or_alias) in user_map {
+                    for id_or_alias in device_id_or_alias {
+                        match id_or_alias {
+                            DeviceIdOrAllDevices::DeviceId(device_id) => {
+                                device_set.insert((user_id.clone(), device_id.clone()));
+                            }
+                            DeviceIdOrAllDevices::AllDevices => {
+                                // get all the devices and add them
+                                let user_devices =
+                                    client.encryption().get_user_devices(&user_id).await?;
+                                let devices = user_devices.devices();
+                                devices.for_each(|device| {
+                                    device_set.insert((
+                                        device.user_id().to_owned(),
+                                        device.device_id().to_owned(),
```

Comment:
This is a bit weird, you are doing the following:

1. Get all the devices of the user
2. Record the user/device id
3. Throw the devices away
4. Iterate through each recorded user/device id and load each device individually from the store.

It seems to me that it would be much better to:

1. Get all the devices of a user
2. If an alias was used, you have your set.
3. If a device IDs were used, filter out any device that isn't part of the recipient set.

Reply:
> I haven't found the simple way of finding failure like you suggested `let failures = recipient_list - recipient_devices`, I tried something else

Sorry, I meant [BTreeSet::difference()](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#method.difference) but couldn't remember the name of the method.